torch/csrc/dynamo/variable_tracker_cache.cpp

```diff
@@ -0,0 +1,224 @@
+#include <torch/csrc/dynamo/variable_tracker_cache.h>
+#include <torch/csrc/utils/python_ptr.h>
+
+#include <functional>
+#include <stdexcept>
+
+namespace {
+
+namespace pyptr = torch::pyptr;
+
+struct VariableTrackerCacheKey {
+  uintptr_t m_value;
+  pyptr::Owned m_source;
+
+  VariableTrackerCacheKey(pyptr::Borrowed value, pyptr::Borrowed source)
+      : m_value(reinterpret_cast<uintptr_t>(value.ptr())),
+        m_source(source.to_owned()) {}
+
+  bool operator==(const VariableTrackerCacheKey& other) const {
+    return (m_value == other.m_value) && m_source.equal(other.m_source);
+  }
+
+  size_t hash() const {
+    return std::hash<uintptr_t>()(m_value) ^ m_source.hash();
+  }
+};
+
+struct VariableTrackerCacheKeyHasher {
+  std::size_t operator()(const VariableTrackerCacheKey& k) const {
+    return k.hash();
+  }
+};
+
+struct VariableTrackerCache {
+  PyObject_HEAD
+
+  std::unordered_map<
+      VariableTrackerCacheKey,
+      pyptr::Owned,
+      VariableTrackerCacheKeyHasher>
+      m_cache;
+
+  ~VariableTrackerCache() {
+    clear();
+  }
+
+  VariableTrackerCache() { // NOLINT(cppcoreguidelines-pro-type-member-init,
+                           // modernize-use-equals-default)
+    // don't use default constructor for objects w/ PyObject_HEAD (crashes).
+    // CLANGTIDY complains about missing init for ob_base.
+  }
+
+  VariableTrackerCache(VariableTrackerCache&) = delete;
+  VariableTrackerCache& operator=(VariableTrackerCache&) = delete;
+  VariableTrackerCache(VariableTrackerCache&&) = delete;
+  VariableTrackerCache& operator=(VariableTrackerCache&&) = delete;
+
+  static PyObject* _new(PyTypeObject* type, PyObject* args, PyObject* kwds) {
+    VariableTrackerCache* self =
+        reinterpret_cast<VariableTrackerCache*>(type->tp_alloc(type, 0));
+    if (self != nullptr) {
+      new (self) VariableTrackerCache();
+    }
+    return reinterpret_cast<PyObject*>(self);
+  }
+
+  static void _dealloc(VariableTrackerCache* self) {
+    self->~VariableTrackerCache();
+    Py_TYPE(self)->tp_free((PyObject*)self);
+  }
+
+  void add(pyptr::Borrowed value, pyptr::Borrowed source, pyptr::Borrowed vt_) {
+    auto vt = vt_.to_owned();
+    auto key = VariableTrackerCacheKey(value, source);
+    m_cache.emplace(std::move(key), vt);
+  }
+
+  void clear() {
+    m_cache.clear();
+  }
+
+  pyptr::Owned clone() {
+    throw std::runtime_error("unimplemented VariableTrackerCache::clone");
+  }
+
+  pyptr::Owned lookup(pyptr::Borrowed value, pyptr::Borrowed source) const {
+    auto it = m_cache.find(VariableTrackerCacheKey(value, source));
```

Comment:
Does constructing this increment the reference count to source?  Possibly that could be avoided.